csrc/codegen.cpp

```diff
@@ -34,6 +34,25 @@ std::string ptrType(DataType dt) {
   return ss.str();
 }
 
+bool isTmaType(const DataType& dtype) {
+  return std::visit(
+      [](auto&& dtype) -> bool {
+        using T = std::decay_t<decltype(dtype)>;
+        if constexpr (std::is_same_v<T, PointerType>) {
+          return isTmaType(*dtype.type);
+        } else if constexpr (std::is_same_v<T, OpaqueType>) {
+          if (0 == dtype.name.compare("TensorMap")) {
+            return true;
+          } else {
+            return false;
+          }
+        } else {
+          return false;
+        }
+      },
```

Comment:
```suggestion
      [](auto&& dtype) -> bool {
        using T = std::decay_t<decltype(dtype)>;
        if constexpr (std::is_same_v<T, PointerType>) {
          return isTmaType(*dtype.type);
        }
        if constexpr (std::is_same_v<T, OpaqueType>) {
          return 0 == dtype.name.compare("TensorMap");
        }
        return false;
      },
```

Reply:
Thanks for suggestion, @wujingyue - you are right, this should be simplified. I will update the code. Thanks!

I made it like this as I wanted to keep it aligned with how `data_type2string(...)` is implemented, but that's an overkill.